src/modules/esl-carousel/renderers/esl-carousel.default.renderer.ts

```diff
@@ -0,0 +1,219 @@
+  }
+
+  /** Makes pre-processing the transition animation of one slide. */
+  protected async onBeforeStepAnimate(direction: ESLCarouselDirection): Promise<void> {
+    this.reorder(this.currentIndex, direction === 'prev');
+
+    const offsetIndex = normalizeIndex(this.currentIndex + (direction === 'next' ? 1 : -1), this.size);
+    const offsetFrom = -this.getOffset(this.currentIndex);
+
+    this.setTransformOffset(offsetFrom);
+    const offsetTo = -this.getOffset(offsetIndex);
+
+    this.$carousel.$$attr('animating', true);
+    this.setTransformOffset(offsetTo);
+
+    if (offsetTo === offsetFrom) return;
+    // TODO: still not catches cases with no animation (
+    await promisifyTransition(this.$area, 'transform');
+  }
+
+  /** Makes post-processing the transition animation of one slide. */
+  protected async onAfterStepAnimate(direction: ESLCarouselDirection): Promise<void> {
+    this.currentIndex = direction === 'next' ? this.currentIndex + 1 : this.currentIndex - 1;
+    this.currentIndex = normalizeIndex(this.currentIndex, this.size);
+
+    this.reorder(this.currentIndex);
+
+    this.setTransformOffset(-this.getOffset(this.currentIndex));
+    this.$carousel.$$attr('animating', false);
+    +this.$carousel.offsetLeft;
+  }
+
+  protected _checkNonLoop(offset: number): boolean {
+    // TODO: get rid of this method by revisiting methods below
+    if (this.loop) return true;
+
+    const sign = offset < 0 ? 1 : -1;
+    const count = Math.floor(Math.abs(offset) / (this.slideSize + this.gap));
+    const nextIndex = this.$carousel.activeIndex + count * sign;
+    const currentIndex = normalizeIndex(this.$carousel.activeIndex + count * sign, this.size);
+
+    // check non-loop state
+    if (nextIndex >= this.$carousel.size || nextIndex < 0) return false;
+    // check left border of non-loop state
+    if (offset > 0 && currentIndex - 1 < 0) return false;
+    // check right border of non-loop state
+    return !(offset < 0 && currentIndex + 1 + this.count > this.$carousel.size);
+  }
+
+  /** Handles the slides transition. */
+  public onMove(offset: number): void {
+    this.$carousel.toggleAttribute('active', true);
+
+    if (!this._checkNonLoop(offset)) return;
+
+    const sign = offset < 0 ? 1 : -1;
+    const slideSize = this.slideSize + this.gap;
+    const count = Math.floor(Math.abs(offset) / slideSize);
+    const currentIndex = normalizeIndex(this.$carousel.activeIndex + count * sign, this.size);
+
+    const orderIndex = offset < 0 ? currentIndex : normalizeIndex(currentIndex - 1, this.size);
+    this.reorder(orderIndex);
+    this.currentIndex = currentIndex;
+
+    const stageOffset = this.getOffset(this.currentIndex) - (offset % slideSize);
+    this.setTransformOffset(-stageOffset);
+  }
+
+  /** Ends current transition and make permanent all changes performed in the transition. */
+  // eslint-disable-next-line sonarjs/cognitive-complexity
+  public async commit(offset: number): Promise<void> {
```

Comment:
for me this is a little bit complicated

Reply:
Agree; I will work on a clearer version with the free-drag mode just after that PR.